### PR TITLE
Fix validation rules for filing, add a could file detection for if th…

### PR DESCRIPTION
**Pull request by BrandonSharratt**
…e user is allowed to do something to a legal type, fix flake violations in continuation_in

*Issue #:* /bcgov/entity#22051

*Description of changes:*
Fixed differences between api and spreadsheet (see below). Also added a could file which just returns if a user could do a filing on a business of the type to the get endpoint.

Differences
Staff
Amalgamation - Regular/Horizontal/Vertical — were additionally blocked on pending filings
Put back on was allowed when it had 'amalgamationApplication' or ’administrative' filings or a future effective dissolution filing
Full Restoration was allowed when there was a continuation in filing
Limited Restoration was allowed when there was a continuation in filing
Change of Registration - was blocked for staff when required info was missing and shouldn’t be
Continuation In didn’t support any of the types in the spreadsheet
Correction required no missing info always not just on SP/GP
Voluntary Dissolution required no missing info always not just on SP/GP
Administrative Dissolution required no missing info always not just on SP/GP
Notice of withdrawal was allowed when business was frozen or had a pending filing

General
Amalgamation - Regular/Horizontal/Vertical — were additionally blocked on pending filings
Continuation In didn’t support any of the types in the spreadsheet
Voluntary Dissolution required no missing info always not just on SP/GP

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of the lear license (Apache 2.0).

### CODE REVIEW

File: legal-api/src/legal_api/resources/v2/business/business.py
Comment (vysakh-menon-aot) [inline legal-api/src/legal_api/resources/v2/business/business.py:81]: Since we are changing this now, its better to move to a separate endpoint (`'/<string:identifier>/allowable-actions'`), as this is required only in filings-ui (and api users).

File: legal-api/src/legal_api/services/authz.py
Comment (vysakh-menon-aot) [inline legal-api/src/legal_api/services/authz.py:607]: If you are planning to use `feeCode` values, there is one exception case (a dynamic fee code based on the user action in alteration. if alteration filing is to convert `BC to ULC` then fee code will be `NOALU` (https://github.com/bcgov/lear/blob/main/legal-api/src/legal_api/models/filing.py#L116) and this fee code will not be available in get_fee_code function

File: legal-api/src/legal_api/services/authz.py
Comment (vysakh-menon-aot) [inline legal-api/src/legal_api/services/authz.py:558]: If I compare this function with `get_allowed_filings` the only change you have done here is to remove `business_blocker_check`. can you explain why?

File: legal-api/src/legal_api/resources/v2/business/business.py
Comment (severinbeauvais) [reply legal-api/src/legal_api/resources/v2/business/business.py:81]: AFAIK, only Filings UI uses allowable actions (though I'd like Edit UI to use it in the future).

File: legal-api/src/legal_api/services/authz.py
Comment (BrandonSharratt) [reply legal-api/src/legal_api/services/authz.py:558]: So the intent behind could_file is to have an api answer on whether a user could ever do a type of filing. So while allowedActions says you can/can't do this for this business, couldFile responds that you could do it for a business of this legalType. We intend to use it to show/hide options, IE couldFile --> visibility, allowedActions --> disabled/enabled

File: legal-api/src/legal_api/services/authz.py
Comment (BrandonSharratt) [reply legal-api/src/legal_api/services/authz.py:607]: We aren't intending to use feeCode, I just wanted to keep the structure the same, and while I would have normally added another flag to allowedActions to remove the blockers check I figured this would be an easier review because it can't possibly affect the current use cases as they all use allowedActions

File: legal-api/src/legal_api/resources/v2/business/business.py
Comment (BrandonSharratt) [reply legal-api/src/legal_api/resources/v2/business/business.py:81]: I'd prefer to not move it to another endpoint, as it increases the network verbosity by requiring 2 api calls when we can currently get it in 1 (and both can be ignored with the slim param)

File: legal-api/src/legal_api/services/authz.py
Comment (severinbeauvais) [reply legal-api/src/legal_api/services/authz.py:558]: I think there "could" (pun intended) be other things that determiend whether or not to show/hide an action (eg, current user).

I'd prefer to defer this implementation until a complete list of show/hide conditions is known.

File: legal-api/src/legal_api/resources/v2/business/business.py
Comment (vysakh-menon-aot) [reply legal-api/src/legal_api/resources/v2/business/business.py:81]: The business endpoint is used in many places, which does not care about allowable actions. I agree it adds an additional api call but it reduces the latency for all other calls to business endpoint

File: legal-api/src/legal_api/services/authz.py
Comment (BrandonSharratt) [reply legal-api/src/legal_api/services/authz.py:558]: The current user and the business type is considered

Comment (vysakh-menon-aot) [reply legal-api/src/legal_api/services/authz.py:558]: If you get all the filings a user can file for a business, how do you check if that action is blocked or not?

Comment (BrandonSharratt) [reply legal-api/src/legal_api/services/authz.py:558]: We still use allowedActions and disable/enable the button based on that flag which checks blocked. the Blocked flag is business specific and doesn't tell you whether or not the current user could file a certain type of filing for a different business of the same type. IE they could file this for a BC business (so it's in could file) but they aren't allowed to for this business (so it isn't in allowedActions)

Comment (argush3) [reply legal-api/src/legal_api/services/authz.py:558]: well, if i'm understanding things correctly.  it doesn't sound like the functionality you are looking for fits under the GET /businesses/<identifier> endpoint.  

It sounds like you need another endpoint that retrieves allowed filings based off of current user and business type.  

So I don't think it makes sense that you are trying to retrieve this info from a specific business?

File: legal-api/src/legal_api/resources/v2/business/business.py
Comment (vysakh-menon-aot) [reply legal-api/src/legal_api/resources/v2/business/business.py:81]: I am not suggesting to remove `business.allowable_actions` from the business endpoint for now, instead introduce new endpoint for the new UI and based on the release of the new UI we can depreciate `business.allowable_actions` in business endpoint.

File: legal-api/src/legal_api/services/authz.py
Comment (severinbeauvais) [reply legal-api/src/legal_api/services/authz.py:558]: > The current user and the business type is considered

The current implementation is simplistic (basically, on or off depending on product key). I know that some user types will have more or less permissions than others (ie, staff > Service BC > reg user).

Before implementing get_could_files, I'd like to see what conditions (in the old UI) are used for showing/hiding actions, and making sure they're considered here.

Comment (severinbeauvais) [reply legal-api/src/legal_api/services/authz.py:558]: I guess, in a nutshell, I'd like to see this feature generalized.

**Comment by BrandonSharratt** [comment]
This PR has been updated and no longer has the DO NOT MERGE tag necessary (that was only needed when the rules were in flight). A test has been added for the new endpoint. Api Documentation/Spec for this new endpoint is to be created after merge as a tech debt ticket

File: legal-api/src/legal_api/resources/v2/business/business.py
Comment (argush3) [inline legal-api/src/legal_api/resources/v2/business/business.py:225]: I would validate to ensure valid business state and business types are being provided.

**Review by argush3 — APPROVED** [review]
Looks good.  

Pre-approving with assumption that business type and business state validation comment will be addressed for new endpoint.

File: legal-api/src/legal_api/resources/v2/business/business.py
Comment (vysakh-menon-aot) [inline legal-api/src/legal_api/resources/v2/business/business.py:227]: it is identifier which starts with T. you can remove this validation

Comment (BrandonSharratt) [reply legal-api/src/legal_api/resources/v2/business/business.py]: Removed

File: legal-api/src/legal_api/resources/v2/business/business.py
Comment (vysakh-menon-aot) [inline legal-api/src/legal_api/resources/v2/business/business.py:230]:
```suggestion
        return {'message': babel('Invalid business type or state.')}, HTTPStatus.BAD_REQUEST
```

**Comment by vysakh-menon-aot** [comment]
Please fix the sonarcloud issue

**Review by severinbeauvais — APPROVED** [review]
LGTM but please update HTTP response code per Vysakh's comment.

File: legal-api/src/legal_api/services/authz.py
Comment (severinbeauvais) [reply legal-api/src/legal_api/services/authz.py]: Obsolete comment. Resolving.

File: legal-api/src/legal_api/resources/v2/business/business.py
Comment (severinbeauvais) [reply legal-api/src/legal_api/resources/v2/business/business.py]: @vysakh-menon-aot Why should this validation be removed?

Comment (vysakh-menon-aot) [reply legal-api/src/legal_api/resources/v2/business/business.py]: this endpoint doesn't have identifier as parameter

**Review by vysakh-menon-aot — APPROVED** [review]

File: legal-api/src/legal_api/resources/v2/business/business.py
Comment (severinbeauvais) [reply legal-api/src/legal_api/resources/v2/business/business.py]: Oh, right 😁 Thanks.

**Pull request merged**
